### PR TITLE
Abstract out RemoteFsTranslogWithPinnedTimestamps

### PR DESCRIPTION
**Pull request by sachinpkale**
### Description
- Abstract out `RemoteFsTranslogWithPinnedTimestamps` from `RemoteFsTranslog` for better separation of concerns.
- https://github.com/opensearch-project/OpenSearch/pull/15416#discussion_r1739640082

### Check List
- [x] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/test/java/org/opensearch/index/translog/RemoteFsTranslogTests.java [hunk 1]
```diff
@@ -671,13 +679,13 @@ public void testSimpleOperationsUpload() throws Exception {
             assertThat(snapshot.totalOperations(), equalTo(ops.size()));
         }
 
-        assertEquals(2, translog.allUploaded().size());
+        assertEquals(4, translog.allUploaded().size());
```
Comment (sachinpkale) [inline server/src/test/java/org/opensearch/index/translog/RemoteFsTranslogTests.java:682]: These changes are reverting back to previous values. We need to change this for combined class with pinned timestamp. As we are abstracting out implementation for pinned timestamp, reverting these back to original values.

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTranslogWithPinnedTimestamps.java [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+
+/**
+ * A Translog implementation which syncs local FS with a remote store
+ * The current impl uploads translog , ckp and metadata to remote store
+ * for every sync, post syncing to disk. Post that, a new generation is
+ * created.
+ *
+ * @opensearch.internal
+ */
+public class RemoteFsTranslogWithPinnedTimestamps extends RemoteFsTranslog {
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/index/translog/RemoteFsTranslogWithPinnedTimestamps.java:52]: nit: rename to `RemoteFsTimestampAwareTranslog`

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTranslogWithPinnedTimestamps.java [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+                        metadataFilesToBeDeleted.remove(metadataFiles.get(0));
+                    }
+
+                    if (metadataFilesToBeDeleted.isEmpty()) {
+                        logger.debug("No metadata files to delete");
+                        remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                        return;
+                    }
+
+                    logger.debug("metadataFilesToBeDeleted = {}", metadataFilesToBeDeleted);
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/index/translog/RemoteFsTranslogWithPinnedTimestamps.java:182]: nit: use lazy loggers?

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTranslogWithPinnedTimestamps.java [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+                } catch (Exception e) {
+                    remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                remoteGenerationDeletionPermits.release(REMOTE_DELETION_PERMITS);
+                logger.error("Exception while listing translog metadata files", e);
+            }
```
Comment (Bukhtawar) [inline server/src/main/java/org/opensearch/index/translog/RemoteFsTranslogWithPinnedTimestamps.java:226]: good to add assertions on all permits

**Review by Bukhtawar — COMMENTED** [review]
Thanks for making these changes, looks much simpler

**Review by Bukhtawar — APPROVED** [review]

File: server/src/main/java/org/opensearch/index/translog/RemoteFsTranslogWithPinnedTimestamps.java (hunk 4 above)
Comment (sachinpkale) [reply server/src/main/java/org/opensearch/index/translog/RemoteFsTranslogWithPinnedTimestamps.java]: The problem is these permits will only be released post generations and metadata delete. So, we won't be able to check if the permits in the end are available or not.

**Review by gbbafna — APPROVED** [review]

**Review by Bukhtawar — APPROVED** [review]
Thanks for the changes Sachin

**Pull request merged**